optimum/onnxruntime/modeling_ort.py

```diff
@@ -346,25 +347,24 @@ def __init__(self, model=None, config=None, **kwargs):
             checkpoint="optimum/all-MiniLM-L6-v2",
         )
     )
-    def forward(
-        self,
-        input_ids: Optional[torch.Tensor] = None,
-        attention_mask: Optional[torch.Tensor] = None,
-        token_type_ids: Optional[torch.Tensor] = None,
-        **kwargs,
-    ):
+    def forward(self, **kwargs):
         # converts pytorch inputs into numpy inputs for onnx
-        onnx_inputs = {
-            "input_ids": input_ids.cpu().detach().numpy(),
-            "attention_mask": attention_mask.cpu().detach().numpy(),
-        }
-        if token_type_ids is not None:
-            onnx_inputs["token_type_ids"] = token_type_ids.cpu().detach().numpy()
+        onnx_inputs = dict(
+            map(lambda input_name: (input_name, kwargs.pop(input_name).cpu().detach().numpy()), self.model_inputs)
+        )
         # run inference
-        outputs = self.model.run(None, onnx_inputs)
-        last_hidden_state = torch.from_numpy(outputs[self.model_outputs["last_hidden_state"]]).to(self.device)
+        onnx_outputs = self.model.run(None, onnx_inputs)
+        outputs = dict(
+            map(
+                lambda output_name: (
+                    output_name,
+                    torch.from_numpy(onnx_outputs[self.model_outputs[output_name]]).to(self.device),
+                ),
+                self.model_outputs,
+            )
+        )
```

Comment:
Have you measured the performance difference for this? Also is there way to improve the readability of the code? having nested `dict`, `map` etc. is not so nice.

Reply:
IMO what users could probably have as inputs for models in transformers, they shall also be able to have them in the `forward` pass of ORTModels. E.g. for BERT model, we could have [various inputs](https://github.com/huggingface/transformers/blob/main/src/transformers/models/bert/modeling_bert.py#L916), but ORTModels only take `input_ids` and `attention_mask` as default and `token_type_ids` as optional. Although this corresponds to the vanilla definition in `FeatureManager`, we shall enable users to take all possible options in transformers as inputs, in case they have exported their model with customized OnnxConfig.
And this is what I did in the `ORTTrainer.evaluate`, I created [wrappers](https://github.com/huggingface/optimum/blob/main/optimum/onnx/configuration.py#L36) to include labels as input and loss as output. Thus the onnx model for evaluation will have labels and loss, and I don't think that could be possible to use pipeline or evaluate or ORTModel to calculate the extra output in such a situation with a a bare onnx model, as we lost information
```
>>> ort_model = ORTModelForSequenceClassification.from_pretrained(model_id="./assets", file_name="distilbert-base-uncased-wo-labels.onnx")
config.json NOT FOUND in HuggingFace Hub
>>> ort_model.config
PretrainedConfig {
  "transformers_version": "4.20.1"
}
```

Reply:
This refactoring does pretty much the same thing as before just in a more abstract way, as the inputs and outputs are defined once we instantiate the ORTModels. 
I do agree that the mapping is not good as one could put no matter what in the `**kwargs` and iterating the invalid inputs could take computation. What about applying a filter for valid keys?

Reply:
> IMO what users could probably have as inputs for models in transformers, they shall also be able to have them in the forward pass of ORTModels. 

I don't agree with that. The `ORTModelForXXX` are created for inference. To provide an easy abstraction from `transformers` to use onnxruntime for inference but keep the same API. 

But before we decide on any changes can you benchmark before and after. Since a `dict(map())` can cost quite a lot,  especially for generation models where the forward pass ist called multiple times.

Reply:
For benchmarking, with the following snippet:
```python
tokens = tokenizer(["I love burritos and pizza and pasta!"]*10000, return_tensors="pt")
start_time = time.time()
pred = ort_model(**tokens)
runtime = time.time() - start_time
result = {"Inference_runtime": round(runtime, 4)}
print(result)
```
Previous ORTModel takes `{'Inference_runtime': 16.6773}` and the current one takes `{'Inference_runtime': 16.6948}`. 

I am agreed that `dict(map())` is not the best option in terms of computation speed. But I think making the ORTModels able to take various valid inputs is something that we should enable in terms of flexibility. What about putting back arguments and adding some `if` control flow as what is done with `token_type_ids`?